Improved code review guidelines (#13294)
opf/openproject: docs/development/code-review-guidelines/README.md

If a review is left untouched, feel free to request a review from a group or link the pull request in question in the devs element channel.

## Taking a review

If you're ready to perform a review for a pull request, do these things:

- Remove the `needs review` label. This is a semaphore and ensures only one developer performs a review
- If there is a linked ticket that has an appropriate status workflow, set it to `in review` and assign yourself
- Optionally, request yourself as a reviewer

## Correctness

As a reviewer, your job is not to make sure that the code is what you would have written – *because it will not be*. Your job as a reviewer of a piece of code is to make sure that the code as written by its author is correct.

Try to think of edge cases when testing or evaluating the code, double check the test coverage. But do not frown if you merged the pull request and something broke after all. This is the learning path to avoiding this mistake on the next attempt. Not doing a review in the first place will not move you forward either.

## Coding style

> We try to adhere to the [Ruby community styleguide](https://github.com/bbatsov/ruby-style-guide) as well as the [AirBnB JavaScript style guide](https://github.com/airbnb/javascript) with [some extensions for Angular](../style-guide/frontend/). Rules we want to follow are expressed as either Rubocop definitions are eslint rules. Follow your linter to adhere to them.

Reviewer: ```suggestion
We try to adhere to the [Ruby community styleguide](https://github.com/bbatsov/ruby-style-guide) as well as the [AirBnB JavaScript style guide](https://github.com/airbnb/javascript) with [some extensions for Angular](../style-guide/frontend/). Rules we want to follow are expressed as either Rubocop definitions or eslint rules. Follow your linter to adhere to them.
```

Due to the age of our codebase, a lot of our code might not yet adhere to these styleguides, but we want all new code to adhere to it. You do not have to improve existing code when making changes, but we encourage it. If you do, please do all improvements in a separate commit from the actual change, so the improvements do not hide your actual code changes in a diff.

Before committing, please run your new code through [Rubocop](https://github.com/bbatsov/rubocop). It detects deviations from a lot of things in the style guide and things that are bad practice in general. You obviously do not have to fix issues with existing code. There is a [list of editor plugins](https://docs.rubocop.org/rubocop/1.31/integration_with_other_tools.html#editor-integration) in the Rubocop docs. You can also use `bin/dirty-rubocop` to test them. Pull requests are being linted automatically through a GitHub action.

> The same is true for eslint. Your editor will likely have support for eslint checks, and allows you to correct them before committing.

Reviewer: Maybe suggest that we have lefthook available and it can be used to install the `pre-commit` hook to automatically check your code before you commit?

Author: good point, I thought about that but that somehow got carried away and didn't document it after all. Will fix that

### Structure of commit messages

- First line: less than 72 characters, this is when GitHub shows ‘…’
- Blank line
- Detailed description of the change, wrapped to 72 characters so the text is readable in git log

See the [Git Book](https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#Commit-Guidelines).

## Testing

- All GitHub workflow actions must be green on continuous integration
- Appropriate unit and integration tests, e.g. test application logic via rspec tests and its integration with the user interface via Cucumber tests
- Test Frontend code with appropriate feature specs.
- Forms: besides testing for success response, also test whether values are actually saved, e.g. read form or look into database
- Bugfixes: must contain a test which detects the bug they fix to prevent regressions
- Translations: Never use a specific translation string in a test. We might want to change them in the future and do not want to fix tests when we do.

> - Manual testing: When testing a feature or change, check out the code test at least the happy paths according to the specification of the ticket.

Reviewer: Maybe a note that we are aware of flickering specs, actively trying to remove them. Something along

> - We are aware that there are some flickering specs in our codebase, that might fail randomly. We are actively trying to fix those. If you encounter test failures in code that you have not touched, try re-running the specs.

*You should not rush through a code review – but also, you need to do it promptly. Your coworkers are waiting for you.*

## Completing the review

> If you have completed the review, you either have left feedback for the developer. In that case:

Reviewer 2: ```suggestion
Once you've completed the review, you might have left feedback for the developer. In that case:
```
